README.md
# DB5 Live

A version of DB5 that supports live updating of values from the plist while running the app on the simulator. Now you can make dozens of tweaks to your app in a matter of minutes!

See the demo app for an example:

1. Run the demo app
2. Change a value in the plist, and save it
3. Watch the app change before your eyes!

Note that this will not work on device.

To add it to your app:

1. Restructure your app delegate (or wherever you load VSThemes in your app) so that you can "refresh" your UI.
2. From this method, call `[themeLoader loadThemes];`
3. Register for file changes by setting the themeReloadedCallback block on themeLoader and have it refresh your UI.
4. Add a `Run Script` build step to your target:
    . $SRCROOT/../Source/VSSymlinkThemesScript.sh
See the demo app for an example, just look at the App Delegate

How it works:
After your app builds, the build script deletes your DB5.plist, and replaces it with a symlink to the source copy. On the simulator, this allows the app to load the updated file without rebuilding.

The code adds the additional magic of watching the file for changes. If the file changes, the callback gets called, which typically will cause your app to refresh. 

For more advanced usage, you could have this callback fire an NSNotification, so that individual View Controllers could reload themselves, rather than reloading the root view controller. This helps when working on a more complex app, as the feedback loop between change and result will be reduced to nearly zero.

## Other Neat Things

* A `CGRect`, `CGSize`, or a `CGPoint` can be defined using a string with comma syntax (ex. `0,0,30,20`)
* `UIColor` can be written in RGB and RGBA using commas, in 255 format (ex. `255,0,0`)
* `UIColor` has an optional alpha component when written in hex

# Original README
# DB5

by [Q Branch](http://qbranch.co/)

## App Configuration via Plist

By storing colors, fonts, numbers, booleans, and so on in a plist, we were able to iterate quickly on our app [Vesper](http://vesperapp.co/).

Our designers could easily make changes without having to dive into the code or ask engineering to spend time nudging pixels and changing values.

There is nothing magical about the code or the system: it’s some simple code plus a few conventions.

### How it works

See the demo app. You include two classes — `VSThemeLoader` and `VSTheme` — and DB5.plist. The plist is where you set values.

At startup you load the file via `VSThemeLoader`, then access values via methods in `VSTheme`.

#### VSTheme methods

Most of the methods are straightforward. `-[VSTheme boolForKey:]` returns a BOOL, and so on.

Some of the methods require multiple values in the plist file. For instance, `-[VSTheme fontForKey:]` expects the font name as `keyName` and the size as `keyNameSize`. See VSTheme.h for more information about these multiple-key values.

#### Inheritance

Though we haven’t used this capability in Vesper, we made it so that you can have multiple themes. Every theme inherits from the Default theme.

If you ask for a value from a theme other than Default, and that value is not specified in that theme, it falls back to Default to get that value.

### Demo app

The demo app is straightforward and small. `DB5AppDelegate` loads the themes. `DB5ViewController` shows some example use.

Also, see the Examples folder for the DB5.plist from Vesper.

### Contact

[Brent Simmons](https://github.com/brentsimmons)<br />
[@brentsimmons](https://twitter.com/brentsimmons)

### License

DB5 is available under the MIT license. See the LICENSE file for details.


CHANGELOG.md
# DB5-Live CHANGELOG

## 0.1.0

Initial release.


Source/VSTheme.h
//
//  VSTheme.h
//  Q Branch LLC
//
//  Created by Brent Simmons on 6/26/13.
//  Copyright (c) 2012 Q Branch LLC. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

typedef NS_ENUM(NSUInteger, VSTextCaseTransform) {
    VSTextCaseTransformNone,
    VSTextCaseTransformUpper,
    VSTextCaseTransformLower
};


@class VSAnimationSpecifier;

@interface VSTheme : NSObject

- (id)initWithDictionary:(NSDictionary *)themeDictionary;

@property (nonatomic, strong) NSString *name;
@property (nonatomic, weak) VSTheme *parentTheme; /*can inherit*/

- (BOOL)boolForKey:(NSString *)key;
- (NSString *)stringForKey:(NSString *)key;
- (NSInteger)integerForKey:(NSString *)key;
- (CGFloat)floatForKey:(NSString *)key;
- (UIImage *)imageForKey:(NSString *)key; /*Via UIImage imageNamed:*/
- (UIColor *)colorForKey:(NSString *)key; /*123ABC or #123ABC: 6 digits, leading # allowed but not required*/
- (UIEdgeInsets)edgeInsetsForKey:(NSString *)key; /*xTop, xLeft, xRight, xBottom keys*/
- (UIFont *)fontForKey:(NSString *)key; /*x and xSize keys*/
- (CGPoint)pointForKey:(NSString *)key; /*xX and xY keys*/
- (CGSize)sizeForKey:(NSString *)key;
- (CGRect)rectForKey:(NSString *)key;
/*xWidth and xHeight keys*/
- (NSTimeInterval)timeIntervalForKey:(NSString *)key;

- (UIViewAnimationOptions)curveForKey:(NSString *)key; /*Possible values: easeinout, easeout, easein, linear*/
- (VSAnimationSpecifier *)animationSpecifierForKey:(NSString *)key; /*xDuration, xDelay, xCurve*/

- (VSTextCaseTransform)textCaseTransformForKey:(NSString *)key; /*lowercase or uppercase -- returns VSTextCaseTransformNone*/

@end


@interface VSTheme (Animations)

- (void)animateWithAnimationSpecifierKey:(NSString *)animationSpecifierKey animations:(void (^)(void))animations completion:(void (^)(BOOL finished))completion;

@end


@interface VSAnimationSpecifier : NSObject

@property (nonatomic, assign) NSTimeInterval delay;
@property (nonatomic, assign) NSTimeInterval duration;
@property (nonatomic, assign) UIViewAnimationOptions curve;

@end



Source/VSThemeLoader.m
//
//  VSThemeLoader.m
//  Q Branch LLC
//
//  Created by Brent Simmons on 6/26/13.
//  Copyright (c) 2012 Q Branch LLC. All rights reserved.
//

#import "VSThemeLoader.h"
#import "VSTheme.h"


@interface VSThemeLoader ()

@property (nonatomic, strong, readwrite) VSTheme *defaultTheme;
@property (nonatomic, strong, readwrite) NSArray *themes;
@property (nonatomic) NSTimeInterval lastThemeModificationDate;
@property (nonatomic, copy) NSString *fileName;
@end


@implementation VSThemeLoader

+ (VSThemeLoader *)sharedThemeLoader
{
    static VSThemeLoader *_instance = nil;

    @synchronized (self) {
        if (_instance == nil) {
            _instance = [[self alloc] init];
        }
    }

    return _instance;
}

- (id)init {
    return [self initWithFileName:@"DB5"];
}

- (id)initWithFileName:(NSString*)fileName {
	self = [super init];
	if (self == nil)
		return nil;

    self.fileName = fileName;
    self.lastThemeModificationDate = [[NSDate date] timeIntervalSinceReferenceDate];
    [self loadThemes];
    
#if TARGET_IPHONE_SIMULATOR
    [NSTimer scheduledTimerWithTimeInterval:.3 target:self selector:@selector(pollFileSystem:) userInfo:nil repeats:YES];
#endif
	return self;
}

- (void)loadThemes
{
    NSString *themesFilePath = [[NSBundle mainBundle] pathForResource:self.fileName ofType:@"plist"];
	NSDictionary *themesDictionary = [NSDictionary dictionaryWithContentsOfFile:themesFilePath];
	
	NSMutableArray *themes = [NSMutableArray array];
	for (NSString *oneKey in themesDictionary) {
		
		VSTheme *theme = [[VSTheme alloc] initWithDictionary:themesDictionary[oneKey]];
		if ([[oneKey lowercaseString] isEqualToString:@"default"])
			_defaultTheme = theme;
		theme.name = oneKey;
		[themes addObject:theme];
	}
    
    for (VSTheme *oneTheme in themes) { /*All themes inherit from the default theme.*/
		if (oneTheme != _defaultTheme)
			oneTheme.parentTheme = _defaultTheme;
    }
    
	_themes = themes;

}

- (VSTheme *)themeNamed:(NSString *)themeName {

	for (VSTheme *oneTheme in self.themes) {
		if ([themeName isEqualToString:oneTheme.name])
			return oneTheme;
	}

	return nil;
}

-(void)pollFileSystem:(NSTimer*)timer
{
#if TARGET_IPHONE_SIMULATOR
    NSString *symlinkPath =
    [[NSFileManager defaultManager]
     destinationOfSymbolicLinkAtPath:[[NSBundle mainBundle] pathForResource:@"DB5" ofType:@"plist" inDirectory:nil]
     error:NULL];
    NSDictionary* fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:symlinkPath error:nil];
    NSDate* modifiedDate = [fileAttributes objectForKey:NSFileModificationDate];
    NSTimeInterval modifiedTimeInterval = [modifiedDate timeIntervalSinceReferenceDate];
    if(self.lastThemeModificationDate > 0 && [modifiedDate timeIntervalSinceReferenceDate] > self.lastThemeModificationDate)
    {
        [self loadThemes];
        self.themeReloadedCallback();
        NSLog(@"Themes Refreshed");
    }
    self.lastThemeModificationDate = modifiedTimeInterval;
#endif
    
}


@end



Demo/DB5Demo/DB5AppDelegate.m
//
//  DB5AppDelegate.m
//  DB5Demo
//
//  Created by Brent Simmons on 6/26/13.
//  Copyright (c) 2013 Q Branch LLC. All rights reserved.
//

#import "DB5AppDelegate.h"
#import "DB5ViewController.h"
#import "VSThemeLoader.h"
#import "VSTheme.h"


@interface DB5AppDelegate ()

@property (nonatomic, strong) VSThemeLoader *themeLoader;

@end


@implementation DB5AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
	

    self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
    self.themeLoader = [VSThemeLoader new];
    __weak DB5AppDelegate* wself = self;
    self.themeLoader.themeReloadedCallback = ^{[wself reloadViewController];};
    [self reloadViewController];
    [self.window makeKeyAndVisible];
    
    
    return YES;
}

- (void)reloadViewController
{
    [self.themeLoader loadThemes];
    self.viewController = [[DB5ViewController alloc] initWithNibName:@"DB5ViewController" bundle:nil theme:self.themeLoader.defaultTheme];
    self.window.rootViewController = self.viewController;
}



@end


Source/VSTheme.m
//
//  VSTheme.m
//  Q Branch LLC
//
//  Created by Brent Simmons on 6/26/13.
//  Copyright (c) 2012 Q Branch LLC. All rights reserved.
//

#import "VSTheme.h"


static BOOL stringIsEmpty(NSString *s);
static UIColor *colorWithHexString(NSString *hexString);


@interface VSTheme ()

@property (nonatomic, strong) NSDictionary *themeDictionary;
@property (nonatomic, strong) NSCache *colorCache;
@property (nonatomic, strong) NSCache *fontCache;

@end


@implementation VSTheme


#pragma mark Init

- (id)initWithDictionary:(NSDictionary *)themeDictionary {
	
	self = [super init];
	if (self == nil)
		return nil;
	
	_themeDictionary = themeDictionary;

	_colorCache = [NSCache new];
	_fontCache = [NSCache new];

	return self;
}


- (id)objectForKey:(NSString *)key {

	id obj = [self.themeDictionary valueForKeyPath:key];
	if (obj == nil && self.parentTheme != nil)
		obj = [self.parentTheme objectForKey:key];
	return obj;
}


- (BOOL)boolForKey:(NSString *)key {

	id obj = [self objectForKey:key];
	if (obj == nil)
		return NO;
	return [obj boolValue];
}


- (NSString *)stringForKey:(NSString *)key {
	
	id obj = [self objectForKey:key];
	if (obj == nil)
		return nil;
	if ([obj isKindOfClass:[NSString class]])
		return obj;
	if ([obj isKindOfClass:[NSNumber class]])
		return [obj stringValue];
	return nil;
}


- (NSInteger)integerForKey:(NSString *)key {

	id obj = [self objectForKey:key];
	if (obj == nil)
		return 0;
	return [obj integerValue];
}


- (CGFloat)floatForKey:(NSString *)key {
	
	id obj = [self objectForKey:key];
	if (obj == nil)
		return  0.0f;
	return [obj floatValue];
}


- (NSTimeInterval)timeIntervalForKey:(NSString *)key {

	id obj = [self objectForKey:key];
	if (obj == nil)
		return 0.0;
	return [obj doubleValue];
}


- (UIImage *)imageForKey:(NSString *)key {
	
	NSString *imageName = [self stringForKey:key];
	if (stringIsEmpty(imageName))
		return nil;
	
	return [UIImage imageNamed:imageName];
}


- (UIColor *)colorForKey:(NSString *)key {

	UIColor *cachedColor = [self.colorCache objectForKey:key];
	if (cachedColor != nil)
		return cachedColor;
    
	NSString *colorString = [self stringForKey:key];

    UIColor *color;
    // Support for RGB and RGBA in 255 (spaces ignored)
    if ([colorString rangeOfString:@","].location != NSNotFound) {
        colorString = [colorString stringByReplacingOccurrencesOfString:@" " withString:@""];
        NSArray *colorComponents = [colorString componentsSeparatedByString:@","];
        if ([colorComponents count] == 3 || [colorComponents count] == 4) {
            float red = [colorComponents[0] floatValue] / 255.0f;
            float green = [colorComponents[1] floatValue] / 255.0f;
            float blue = [colorComponents[2] floatValue] / 255.0f;
            float alpha = 1.0;
            if ([colorComponents count] == 4) {
                alpha = [colorComponents[3] floatValue] / 255.0f;
            }
            color = [UIColor colorWithRed:red green:green blue:blue alpha:alpha];
        }
    }
    else
    {
        color = colorWithHexString(colorString);
    }
	if (color == nil)
		color = [UIColor purpleColor];

	[self.colorCache setObject:color forKey:key];

	return color;
}


- (UIEdgeInsets)edgeInsetsForKey:(NSString *)key {

	CGFloat left = [self floatForKey:[key stringByAppendingString:@"Left"]];
	CGFloat top = [self floatForKey:[key stringByAppendingString:@"Top"]];
	CGFloat right = [self floatForKey:[key stringByAppendingString:@"Right"]];
	CGFloat bottom = [self floatForKey:[key stringByAppendingString:@"Bottom"]];

	UIEdgeInsets edgeInsets = UIEdgeInsetsMake(top, left, bottom, right);
	return edgeInsets;
}


- (UIFont *)fontForKey:(NSString *)key {

	UIFont *cachedFont = [self.fontCache objectForKey:key];
	if (cachedFont != nil)
		return cachedFont;
    
	NSString *fontName = [self stringForKey:key];
	CGFloat fontSize = [self floatForKey:[key stringByAppendingString:@"Size"]];

	if (fontSize < 1.0f)
		fontSize = 15.0f;

	UIFont *font = nil;
    
	if (stringIsEmpty(fontName))
		font = [UIFont systemFontOfSize:fontSize];
	else
		font = [UIFont fontWithName:fontName size:fontSize];

	if (font == nil)
		font = [UIFont systemFontOfSize:fontSize];
    
	[self.fontCache setObject:font forKey:key];

	return font;
}


- (CGPoint)pointForKey:(NSString *)key {
    NSString *pointString = [self stringForKey:key];
    if (pointString && [pointString rangeOfString:@","].location != NSNotFound) {
        NSArray *pointComponents = [pointString componentsSeparatedByString:@","];
        if ([pointComponents count] == 2) {
            CGFloat x = [pointComponents[0] floatValue];
            CGFloat y = [pointComponents[1] floatValue];
            return CGPointMake(x, y);
        }
        NSLog(@"DB5: Unable to find point for key %@", key);
        return CGPointZero;
    }
	CGFloat pointX = [self floatForKey:[key stringByAppendingString:@"X"]];
	CGFloat pointY = [self floatForKey:[key stringByAppendingString:@"Y"]];

	CGPoint point = CGPointMake(pointX, pointY);
	return point;
}

- (CGSize)sizeForKey:(NSString *)key {
    NSString *sizeString = [self stringForKey:key];
    if (sizeString && [sizeString rangeOfString:@","].location != NSNotFound) {
        NSArray *sizeComponents = [sizeString componentsSeparatedByString:@","];
        if ([sizeComponents count] == 2) {
            CGFloat x = [sizeComponents[0] floatValue];
            CGFloat y = [sizeComponents[1] floatValue];
            return CGSizeMake(x, y);
        }
        NSLog(@"DB5: Unable to find size for key %@", key);
        return CGSizeZero;
    }
    CGFloat width = [self floatForKey:[key stringByAppendingString:@"Width"]];
    CGFloat height = [self floatForKey:[key stringByAppendingString:@"Height"]];
    CGSize size = CGSizeMake(width, height);
    return size;
}

- (CGRect)rectForKey:(NSString*)key {
    NSString* rectString = [self stringForKey:key];
    if (rectString && [rectString rangeOfString:@","].location != NSNotFound) {
        NSArray *rectComponents = [rectString componentsSeparatedByString:@","];
        if ([rectComponents count] == 4) {
            CGFloat x = [rectComponents[0] floatValue];
            CGFloat y = [rectComponents[1] floatValue];
            CGFloat w = [rectComponents[2] floatValue];
            CGFloat h = [rectComponents[3] floatValue];
            return CGRectMake(x, y, w, h);
        }
        NSLog(@"DB5: Unable to find rect for key %@", key);
        return CGRectZero;
    }
    return (CGRect){[self pointForKey:key], [self sizeForKey:key]};

}

- (UIViewAnimationOptions)curveForKey:(NSString *)key {
    
	NSString *curveString = [self stringForKey:key];
	if (stringIsEmpty(curveString))
		return UIViewAnimationOptionCurveEaseInOut;

	curveString = [curveString lowercaseString];
	if ([curveString isEqualToString:@"easeinout"])
		return UIViewAnimationOptionCurveEaseInOut;
	else if ([curveString isEqualToString:@"easeout"])
		return UIViewAnimationOptionCurveEaseOut;
	else if ([curveString isEqualToString:@"easein"])
		return UIViewAnimationOptionCurveEaseIn;
	else if ([curveString isEqualToString:@"linear"])
		return UIViewAnimationOptionCurveLinear;
    
	return UIViewAnimationOptionCurveEaseInOut;
}


- (VSAnimationSpecifier *)animationSpecifierForKey:(NSString *)key {

	VSAnimationSpecifier *animationSpecifier = [VSAnimationSpecifier new];

	animationSpecifier.duration = [self timeIntervalForKey:[key stringByAppendingString:@"Duration"]];
	animationSpecifier.delay = [self timeIntervalForKey:[key stringByAppendingString:@"Delay"]];
	animationSpecifier.curve = [self curveForKey:[key stringByAppendingString:@"Curve"]];

	return animationSpecifier;
}


- (VSTextCaseTransform)textCaseTransformForKey:(NSString *)key {

	NSString *s = [self stringForKey:key];
	if (s == nil)
		return VSTextCaseTransformNone;

	if ([s caseInsensitiveCompare:@"lowercase"] == NSOrderedSame)
		return VSTextCaseTransformLower;
	else if ([s caseInsensitiveCompare:@"uppercase"] == NSOrderedSame)
		return VSTextCaseTransformUpper;

	return VSTextCaseTransformNone;
}


@end


@implementation VSTheme (Animations)


- (void)animateWithAnimationSpecifierKey:(NSString *)animationSpecifierKey animations:(void (^)(void))animations completion:(void (^)(BOOL finished))completion {

    VSAnimationSpecifier *animationSpecifier = [self animationSpecifierForKey:animationSpecifierKey];

    [UIView animateWithDuration:animationSpecifier.duration delay:animationSpecifier.delay options:animationSpecifier.curve animations:animations completion:completion];
}

@end


#pragma mark -

@implementation VSAnimationSpecifier

@end


static BOOL stringIsEmpty(NSString *s) {
	return s == nil || [s length] == 0;
}


static UIColor *colorWithHexString(NSString *hexString) {

	/*Picky. Crashes by design.*/
	
	if (stringIsEmpty(hexString))
		return [UIColor blackColor];

	NSMutableString *s = [hexString mutableCopy];
	[s replaceOccurrencesOfString:@"#" withString:@"" options:0 range:NSMakeRange(0, [hexString length])];
	CFStringTrimWhitespace((__bridge CFMutableStringRef)s);

	NSString *redString = [s substringToIndex:2];
	NSString *greenString = [s substringWithRange:NSMakeRange(2, 2)];
	NSString *blueString = [s substringWithRange:NSMakeRange(4, 2)];

	unsigned int red = 0, green = 0, blue = 0, alpha = 255;
	[[NSScanner scannerWithString:redString] scanHexInt:&red];
	[[NSScanner scannerWithString:greenString] scanHexInt:&green];
	[[NSScanner scannerWithString:blueString] scanHexInt:&blue];
    if([s length] == 8)
    {
        NSString *alphaString = [s substringWithRange:NSMakeRange(6, 2)];
        [[NSScanner scannerWithString:alphaString] scanHexInt:&alpha];
    }

	return [UIColor colorWithRed:(CGFloat)red/255.0f green:(CGFloat)green/255.0f blue:(CGFloat)blue/255.0f alpha:alpha/255.0f];
}
